lib/IDE/Refactoring.cpp

```diff
@@ -1448,6 +1448,98 @@ bool RefactoringActionExtractRepeatedExpr::performChange() {
                                           EditConsumer).performChange();
 }
 
+static std::unique_ptr<llvm::SetVector<Expr*>> findConcatenatedExpressions(ResolvedRangeInfo Info) {
+  if (Info.Kind != RangeKind::SingleExpression && Info.Kind != RangeKind::PartOfExpression)
+    return nullptr;
+  Expr *E = Info.ContainedNodes[0].get<Expr*>();
+
+  struct StringInterpolationExprFinder: public SourceEntityWalker {
+    std::unique_ptr<llvm::SetVector<Expr*>> Bucket = llvm::make_unique<llvm::SetVector<Expr*>>();
+    bool IsValidInterpolation = true;
+    StringInterpolationExprFinder() {}
+
+    bool isConcatenationExpr(DeclRefExpr* Expr) {
+      if (!Expr)
+        return false;
+      auto *FD = dyn_cast<FuncDecl>(Expr->getDecl());
+      if (FD == nullptr || FD->getName().str() != "+") {
+        return false;
+      }
+      return true;
+    }
+
+    bool walkToExprPre(Expr *E) {
+      if (E->isImplicit())
+        return true;
+
+      auto ExprType = E->getType().getString();
+      //Only binary concatenation operators should exist in expression
+      if (E->getKind() == ExprKind::Binary) {
+        auto *BE = dyn_cast<BinaryExpr>(E);
+        auto *OperatorDeclRef = BE->getSemanticFn()->getMemberOperatorRef();
+        if (!isConcatenationExpr(OperatorDeclRef) || ExprType != "String") {
+          IsValidInterpolation = false;
+          return false;
+        }
+        return true;
+      }
+      //Everything that evaluates to string should be gathered
+      if (ExprType == "String") {
+        Bucket->insert(E);
+        return false;
+      }
+      if (E->getKind() == ExprKind::DeclRef) {
+        auto *DR = dyn_cast<DeclRefExpr>(E);
+        auto *FD = dyn_cast<FuncDecl>(DR->getDecl());
+        auto IsConcatenation = isConcatenationExpr(DR);
+        if (FD && IsConcatenation) {
+          return false;
+        }
+      }
+      //There was non-expected expression, it's not valid interpolation then
+      IsValidInterpolation = false;
+      return false;
+    }
+  } Walker;
+  Walker.walk(E);
+  //There should be two or more expressions to convert
+  if (!Walker.IsValidInterpolation || Walker.Bucket->size() < 2)
+    return nullptr;
+  return std::move(Walker.Bucket);
```

Comment:
We don't need `std::move` here.

Reply:
Hm, I'm returning `std::unique_ptr` which has deleted copy constructor, so I need to move the membership if the func signature is `static std::unique_ptr<llvm::SetVector<Expr*>>
  findConcatenatedExpressions(ResolvedRangeInfo Info)`. Am I missing something?

Reply:
oh, you are right. We don't need `std::move` if the unique pointer is a local variable.

Reply:
Great, thanks!